src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,956 @@
+
+            public override BoundNode? VisitTypePattern(BoundTypePattern node)
+            {
+                return NegateIfNeeded(node);
+            }
+
+            public override BoundNode? VisitConstantPattern(BoundConstantPattern node)
+            {
+                return NegateIfNeeded(node);
+            }
+
+            public override BoundNode? VisitDiscardPattern(BoundDiscardPattern node)
+            {
+                return NegateIfNeeded(node);
+            }
+
+            public override BoundNode? VisitDeclarationPattern(BoundDeclarationPattern node)
+            {
+                BoundPattern result;
+                if (node.IsVar)
+                {
+                    result = MakeDiscardPattern(node).MakeCompilerGenerated();
+                }
+                else
+                {
+                    if (node.InputType.Equals(node.DeclaredType.Type)) // TODO2 what kind of type comparison?
+                    {
+                        result = MakeDiscardPattern(node).MakeCompilerGenerated();
+                    }
+                    else
+                    {
+                        result = node;
+                    }
+                }
+
+                return NegateIfNeeded(result);
+            }
+
+            public override BoundNode? VisitRelationalPattern(BoundRelationalPattern node)
+            {
+                return NegateIfNeeded(node);
+            }
+
+            public override BoundNode? VisitRecursivePattern(BoundRecursivePattern node)
+            {
+                if (!_negated)
+                {
+                    return base.VisitRecursivePattern(node);
+                }
+
+                // `Type (D1, D2, ...) { Prop1: P1, Prop2: P2, ... } x`
+                // can be expanded to `Type and Type (D1, _, ...) and Type (_, D2, ...) and Type { Prop1: P1 } and Type { Prop2: P2 } or ... or var x`
+                // which can be negated to
+                // `not Type or Type (not D1, _, ...) or Type (_, not D2, ...) or Type { Prop1: not P1 } or Type { Prop2: not P2 } or ...`
+                // If there's not Type, we substitute a null check
+
+                var builder = ArrayBuilder<BoundPattern>.GetInstance();
+
+                if (node.DeclaredType is not null)
+                {
+                    // `not Type`
+                    builder.Add(new BoundNegatedPattern(node.Syntax,
+                        new BoundTypePattern(node.Syntax, node.DeclaredType, node.IsExplicitNotNullTest, node.InputType, node.NarrowedType, node.HasErrors),
+                        node.InputType, node.NarrowedType, node.HasErrors).MakeCompilerGenerated());
+                }
+                else if (node.InputType.CanContainNull())
+                {
+                    // `null`
+                    builder.Add(new BoundConstantPattern(node.Syntax,
+                        new BoundLiteral(node.Syntax, constantValueOpt: null, type: null, hasErrors: false), // dummy expression
```

Comment:
>new BoundLiteral(node.Syntax, constantValueOpt: null, type: null, hasErrors: false) [](http://example.com/codeflow?start=24&length=83)

Is this supposed to be a null literal expression?

Reply:
Yes